cdm/core/src/test/java/ucar/nc2/filter/TestEnhancements.java

```diff
@@ -0,0 +1,145 @@
+    writer.write(writer.findVariable("scaleOffsetVar"), new int[1], signedData);
+    writer.write(writer.findVariable("validMin"), new int[1], missingDataArray);
+    writer.write(writer.findVariable("validMinMax"), new int[1], missingDataArray);
+    writer.write(writer.findVariable("validRange"), new int[1], missingDataArray);
+    writer.write(writer.findVariable("enhanceAll"), new int[1], enhanceAllArray);
+    writer.close();
+    ncd = NetcdfDatasets.openDataset(filePath);
+  }
+
+  @Test
+  public void testUnsignedConversion() throws IOException {
+    final int[] unsignedValues = new int[] {123, 124, 125, 126, 127, -128, -127, -126, -125, -126};
+    // signed var
+    Variable v = ncd.findVariable("signedVar");
+    Array data = v.read();
+    assertThat(data.isUnsigned()).isFalse();
+    assertThat(data.getDataType()).isEqualTo(DataType.SHORT);
+    assertThat((short[]) data.copyTo1DJavaArray()).isEqualTo(signedShorts);
+
+    // var with unsigned data type
+    v = ncd.findVariable("unsignedVar");
+    data = v.read();
+    assertThat(data.isUnsigned()).isTrue();
+    assertThat(data.getDataType()).isEqualTo(DataType.UINT);
+    assertThat((int[]) data.copyTo1DJavaArray()).isEqualTo(unsignedValues);
+  }
+
+  @Test
+  public void testScaleOffset() throws IOException {
+    final double[] expected = new double[] {1240, 1250, 1260, 1270, 1280, -1270, -1260, -1250, -1240, -1230};
+    // signed var
+    Variable v = ncd.findVariable("scaleOffsetVar");
+    Array data = v.read();
+    assertThat(data.isUnsigned()).isFalse();
+    assertThat(data.getDataType()).isEqualTo(DataType.SHORT);
+    assertThat((double[]) data.copyTo1DJavaArray()).isEqualTo(expected);
+  }
+
+  @Test
+  public void testConvertMissing() throws IOException {
+    // var with valid min
+    float[] expected = new float[] {Float.NaN, 100, Float.NaN, 120, 130, 140, 150, 190, 200, 210};
+    Variable v = ncd.findVariable("validMin");
+    Array data = v.read();
+    assertThat((float[]) data.copyTo1DJavaArray()).isEqualTo(expected);
+
+    // var with valid min and max
+    expected = new float[] {Float.NaN, 100, Float.NaN, 120, 130, 140, 150, 190, 200, Float.NaN};
+    v = ncd.findVariable("validMinMax");
+    data = v.read();
+    assertThat((float[]) data.copyTo1DJavaArray()).isEqualTo(expected);
+
+    // var with valid range and fill value
+    expected = new float[] {Float.NaN, 100, Float.NaN, 120, 130, 140, Float.NaN, 190, 200, Float.NaN};
+    v = ncd.findVariable("validRange");
+    data = v.read();
+    assertThat((float[]) data.copyTo1DJavaArray()).isEqualTo(expected);
+  }
+
+  @Test
+  public void testCombinedEnhancements() throws IOException {
+    int[] expected = new int[] {1240, 1250, 1260, 1270, 1280, 0, 1278, 1260, 0, 0};
+    Variable v = ncd.findVariable("enhanceAll");
+    Array data = v.read();
+    assertThat((int[]) data.copyTo1DJavaArray()).isEqualTo(expected);
+  }
+}
```

Comment:
One test I'd be curious to see is if I take values that are from -1 to 1 and encode them as unsigned 8 bit integers (so, scale_factor = 0.008 and add_offset = -1), will the decode process give me back an array of floats rather than doubles. It isn't immediately clear to me that it would do that given that internally the scale and offset are stored as doubles.